### PR TITLE
feat: extension for latest commit information 

### PR DESCRIPTION
**Pull request by omarsilva1**
# Pull Request
This PR adds an extension with rest APIs to get the latest commit information of the current jar and image.

## How Has This Been Tested?
Test classes.

## Linked Issue(s)
Closes https://github.com/sovity/EDC-deployments/issues/165

# Checklist
- [x] I have **formatted the title** correctly and precisely
- [x] My code follows the **style guidelines** of this project
- [x] I have performed a **self-review** of my own code
- [x] I have **commented** my code, particularly in hard-to-understand areas and public classes/methods
- [x] I have made corresponding changes to the **documentation**
- [x] My changes generate **no new warnings** (performed checkstyle check locally)
- [x] I have added **tests that prove my fix** is effective or that my feature works
- [x] New and existing unit **tests pass locally** with my changes
- [x] Any dependent **changes have been merged** and published in downstream modules
- [x] I have added/updated **copyright headers**

### CODE REVIEW

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -30,6 +30,15 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v3
 
+      - name: Store last commit's author, message and SHA
+        id: last-commit-information
+        run: |
+          echo "Saving local commit details."
+          echo "LAST_COMMIT_INFORMATION<<EOF" >> $GITHUB_ENV
+          export LAST_COMMIT_INFORMATION=$(git log -1)
+          echo "$LAST_COMMIT_INFORMATION" >> $GITHUB_ENV
```
Comment (richardtreier) [inline .github/workflows/docker-image.yml:39]: can also be shortened to last-commit-info for global naming consistency

Comment (richardtreier) [reply .github/workflows/docker-image.yml:39]: Could be shortened to `git log -1 >> $GITHUB_ENV`, right?

File: connector/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,6 @@
 FROM gradle:7.6.0-jdk17 AS build
 
+ARG EDC_LAST_COMMIT_INFO="The GitHub Pipeline will replace this string during build time with the commit information (SHA, author and commit message) of the built image."
```
Comment (richardtreier) [inline connector/Dockerfile:3]: "The docker container was built outside of github actions and you didn't provide the build arg EDC_LAST_COMMIT_INFO_ARG, so there's no last commit info."

Also should contain a comment above, that we want to se the env var from the github action, but there's no way in the plugin to specify env vars, so we specify a build arg that we'll copy into the env var.

File: connector/Dockerfile [hunk 3]
```diff
@@ -35,6 +36,7 @@ ENV WEB_HTTP_DATA_PORT="9191"
 ENV WEB_HTTP_DATA_PATH="/api/v1/data"
 ENV WEB_HTTP_IDS_PORT="8282"
 ENV WEB_HTTP_IDS_PATH="/api/v1/ids"
+ENV EDC_LAST_COMMIT_INFO_ENV=$EDC_LAST_COMMIT_INFO
```
Comment (richardtreier) [inline connector/Dockerfile:39]: The ENV Var will be adjustable on container startup, via docker compose, etc., so it's the (more) true API.

So it should be the other way around, with the ENV var not having the ENV suffix and the build arg having the ARG suffix and an explanation why it even exists.

File: extensions/last-commit-info/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.8.1")
+}
+
+publishing {
+    publications {
+        create<MavenPublication>(project.name) {
+            from(components["java"])
+        }
+    }
+}
```
Comment (richardtreier) [inline extensions/last-commit-info/build.gradle.kts:33]: missing newline

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                                    LastCommitInfoService lastCommitInfoService) {
+        this.monitor = monitor;
+        this.lastCommitInfoService = lastCommitInfoService;
+    }
+
+    @GET
+    @Path("/")
+    public String getLastCommitInformation() {
+        var result = lastCommitInfoService.getLastCommitInfo();
+        monitor.info(result);
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java:39]: Why do we need the server-side logging? This is potentially visible to customers, too.

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        return result;
+    }
+
+    @GET
+    @Path("/env")
+    public String getLastEnvCommitInformation() {
+        var result = lastCommitInfoService.getEnvLastCommitInfo();
+        monitor.info(result);
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java:47]: same

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    @Path("/")
+    public String getLastCommitInformation() {
+        var result = lastCommitInfoService.getLastCommitInfo();
+        monitor.info(result);
+        return result;
+    }
+
+    @GET
+    @Path("/env")
+    public String getLastEnvCommitInformation() {
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java:45]: We should not need two endpoints. We should have one endpoint that produces text/plain that concatenates the messages.

"last commit info\n\nedc-extensions last commit info:\njar commit info" (second part only if they aren't the same)

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    public String getJarLastCommitInfo() {
+        var classloader = Thread.currentThread().getContextClassLoader();
+        var is = classloader.getResourceAsStream("jar-last-commit-info.txt");
+        var scanner = new Scanner(Objects.requireNonNull(is)).useDelimiter("\\A");
+        return scanner.hasNext() ? scanner.next() : "";
+    }
+
+    public String getEnvLastCommitInfo() {
+        return context.getSetting("edc.last.commit.info.env", "");
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java:46]: also remove .env suffix here

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        result += getEnvLastCommitInfo() + "\n";
+        result += "Jar Last Commit Info: \n";
+        result += getJarLastCommitInfo();
+        return result;
+    }
+
+    public String getJarLastCommitInfo() {
+        var classloader = Thread.currentThread().getContextClassLoader();
+        var is = classloader.getResourceAsStream("jar-last-commit-info.txt");
+        var scanner = new Scanner(Objects.requireNonNull(is)).useDelimiter("\\A");
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java:41]: missing UTF-8 encoding

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    private final ServiceExtensionContext context;
+
+    public LastCommitInfoService(ServiceExtensionContext context) {
+        this.context = context;
+    }
+
+    public String getLastCommitInfo() {
+        var result = "Env Last Commit Info: \n";
+        result += getEnvLastCommitInfo() + "\n";
+        result += "Jar Last Commit Info: \n";
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java:33]: should only be added if it differs from the env last commit info, and if it does, can be safely called "edc-extensions last commit info"

**Review by richardtreier — CHANGES_REQUESTED** [review]

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        result += getJarLastCommitInfo();
+        return result;
+    }
+
+    public String getJarLastCommitInfo() {
+        var classloader = Thread.currentThread().getContextClassLoader();
```
Comment (richardtreier) [inline extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoService.java:39]: classLoader

File: extensions/last-commit-info/src/test/java/de/sovity/edc/extension/version/controller/LastCommitInfoControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+                .header("x-api-key", authKey)
+                .when()
+                .contentType(ContentType.TEXT)
+                .get(String.format("/last-commit-info/env"))
+                .then()
+                .statusCode(200)
+                .contentType(ContentType.JSON);
+        request.assertThat().body(containsString("pipeline"));
+    }
+}
```
Comment (richardtreier) [inline extensions/last-commit-info/src/test/java/de/sovity/edc/extension/version/controller/LastCommitInfoControllerTest.java:77]: newline

File: connector/Dockerfile (hunk 2 above)
Comment (omarsilva1) [reply connector/Dockerfile]: done

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java (hunk 5 above)
Comment (omarsilva1) [reply extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java]: done

File: extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java (hunk 6 above)
Comment (omarsilva1) [reply extensions/last-commit-info/src/main/java/de/sovity/edc/extension/LastCommitInfoController.java]: done

**Review by richardtreier — APPROVED** [review]

**Review by richardtreier — APPROVED** [review]

**Pull request merged**